File: src/utils/dataProcessor.ts
import { SalesRecord, DataProcessingJob, Product } from '../types';
import { SKUMapper } from './skuMapper';

export class DataProcessor {
  private skuMapper: SKUMapper;

  constructor() {
    this.skuMapper = new SKUMapper();
  }

  public async processFile(
    file: File,
    marketplace: string,
    existingProducts: Product[],
    onProgress?: (progress: number) => void
  ): Promise<DataProcessingJob> {
    const job: DataProcessingJob = {
      id: `job_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
      fileName: file.name,
      fileSize: file.size,
      marketplace,
      status: 'processing',
      progress: 0,
      recordsProcessed: 0,
      recordsTotal: 0,
      errors: [],
      warnings: [],
      createdAt: new Date(),
    };

    try {
      const content = await this.readFileContent(file);
      const records = await this.parseCSVContent(content);
      
      job.recordsTotal = records.length;
      
      const processedRecords: SalesRecord[] = [];
      
      for (let i = 0; i < records.length; i++) {
        try {
          const record = await this.processRecord(records[i], marketplace, existingProducts, i + 1);
          if (record) {
            processedRecords.push(record);
          }
          
          job.recordsProcessed = i + 1;
          job.progress = Math.round((i + 1) / records.length * 100);
          
          if (onProgress) {
            onProgress(job.progress);
          }
          
          // Simulate processing delay
          await new Promise(resolve => setTimeout(resolve, 10));
        } catch (error) {
          job.errors.push(`Row ${i + 1}: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
      }

      job.status = 'completed';
      job.completedAt = new Date();
      
      // Store processed records (in a real app, this would go to a database)
      this.storeProcessedRecords(processedRecords);
      
    } catch (error) {
      job.status = 'failed';
      job.errors.push(error instanceof Error ? error.message : 'Unknown error occurred');
    }

    return job;
  }

  private async readFileContent(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = (e) => resolve(e.target?.result as string);
      reader.onerror = () => reject(new Error('Failed to read file'));
      reader.readAsText(file);
    });
  }

  private async parseCSVContent(content: string): Promise<any[]> {
    const lines = content.split('\n').filter(line => line.trim());
    if (lines.length === 0) throw new Error('Empty file');

    // Handle CSV parsing with proper quote handling
    const parseCSVLine = (line: string): string[] => {
      const result: string[] = [];
      let current = '';
      let inQuotes = false;
      
      for (let i = 0; i < line.length; i++) {
        const char = line[i];
        const nextChar = line[i + 1];
        
        if (char === '"') {
          if (inQuotes && nextChar === '"') {
            current += '"';
            i++; // Skip next quote
          } else {
            inQuotes = !inQuotes;
          }
        } else if (char === ',' && !inQuotes) {
          result.push(current.trim());
          current = '';
        } else {
          current += char;
        }
      }
      
      result.push(current.trim());
      return result;
    };

    const headers = parseCSVLine(lines[0]).map(h => h.replace(/"/g, '').trim());
    const records = [];

    for (let i = 1; i < lines.length; i++) {
      const values = parseCSVLine(lines[i]).map(v => v.replace(/^"|"$/g, '').trim());
      const record: any = {};
      
      headers.forEach((header, index) => {
        record[header] = values[index] || '';
      });
      
      // Skip empty rows
      if (Object.values(record).some(value => value && value.toString().trim())) {
        records.push(record);
      }
    }

    return records;
  }

  private async processRecord(
    rawRecord: any,
    marketplace: string,
    existingProducts: Product[],
    rowNumber: number
  ): Promise<SalesRecord | null> {
    // Get comprehensive header mapping for the marketplace
    const headerMapping = this.getHeaderMapping(marketplace);
    const mappedRecord: any = {};

    // Map fields with better matching logic
    for (const [ourField, possibleHeaders] of Object.entries(headerMapping)) {
      let value = null;
      
      // Try exact matches first
      for (const header of possibleHeaders) {
        if (rawRecord[header] !== undefined && rawRecord[header] !== '') {
          value = rawRecord[header];
          break;
        }
      }
      
      // If no exact match, try case-insensitive partial matches
      if (value === null) {
        const recordKeys = Object.keys(rawRecord);
        for (const header of possibleHeaders) {
          const matchingKey = recordKeys.find(key => 
            key.toLowerCase().includes(header.toLowerCase()) ||
            header.toLowerCase().includes(key.toLowerCase())
          );
          if (matchingKey && rawRecord[matchingKey] !== undefined && rawRecord[matchingKey] !== '') {
            value = rawRecord[matchingKey];
            break;
          }
        }
      }
      
      mappedRecord[ourField] = value;
    }

    // Enhanced validation with better error messages
    const validationErrors: string[] = [];
    
    if (!mappedRecord.sku || mappedRecord.sku.toString().trim() === '') {
      // Try to find any field that might be a SKU
      const possibleSkuFields = Object.keys(rawRecord).filter(key => 
        key.toLowerCase().includes('sku') || 
        key.toLowerCase().includes('asin') ||
        key.toLowerCase().includes('item') ||
        key.toLowerCase().includes('product')
      );
      
      if (possibleSkuFields.length > 0) {
        mappedRecord.sku = rawRecord[possibleSkuFields[0]];
      } else {
        validationErrors.push('Missing required field: SKU (also looked for ASIN, Item ID, Product ID)');
      }
    }
    
    if (!mappedRecord.orderId || mappedRecord.orderId.toString().trim() === '') {
      // For Amazon inventory reports, use a combination of Date + FNSKU + ASIN as unique identifier
      if (rawRecord['Date'] && (rawRecord['FNSKU'] || rawRecord['ASIN'])) {
        mappedRecord.orderId = `${rawRecord['Date']}-${rawRecord['FNSKU'] || rawRecord['ASIN']}-${rowNumber}`;
      } else {
        validationErrors.push('Missing required field: Order ID (for inventory reports, using Date + FNSKU/ASIN combination)');
      }
    }

    if (validationErrors.length > 0) {
      throw new Error(validationErrors.join('; '));
    }

    // Process SKU mapping
    const skuMapping = this.skuMapper.mapSKU(mappedRecord.sku, marketplace, existingProducts);

    // Create sales record with better default handling
    const salesRecord: SalesRecord = {
      id: `sale_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
      orderId: mappedRecord.orderId.toString(),
      marketplace,
      sku: skuMapping.originalSku,
      msku: skuMapping.msku,
      productName: mappedRecord.productName || mappedRecord.sku || 'Unknown Product',
      quantity: this.parseNumber(mappedRecord.quantity, 1),
      unitPrice: this.parseNumber(mappedRecord.unitPrice, 0),
      totalAmount: this.parseNumber(mappedRecord.totalAmount, 0),
      fees: this.parseNumber(mappedRecord.fees, 0),
      netAmount: this.parseNumber(mappedRecord.netAmount, 0),
      orderDate: this.parseDate(mappedRecord.orderDate),
      status: mappedRecord.status || 'pending',
    };

    // Calculate derived fields
    if (!salesRecord.totalAmount && salesRecord.unitPrice && salesRecord.quantity) {
      salesRecord.totalAmount = salesRecord.unitPrice * salesRecord.quantity;
    }

    if (!salesRecord.netAmount) {
      salesRecord.netAmount = salesRecord.totalAmount - salesRecord.fees;
    }

    return salesRecord;
  }

  private parseNumber(value: any, defaultValue: number = 0): number {
    if (value === null || value === undefined || value === '') {
      return defaultValue;
    }
    
    // Remove currency symbols and commas
    const cleanValue = value.toString().replace(/[$,£€¥]/g, '').trim();
    const parsed = parseFloat(cleanValue);
    
    return isNaN(parsed) ? defaultValue : parsed;
  }

  private parseDate(value: any): Date {
    if (!value) return new Date();
    
    try {
      const date = new Date(value);
      return isNaN(date.getTime()) ? new Date() : date;
    } catch {
      return new Date();
    }
  }

  private getHeaderMapping(marketplace: string): Record<string, string[]> {
    const baseMapping = {
      orderId: [
        'order-id', 'Order ID', 'OrderId', 'order_id', 'Order Number',
        'amazon-order-id', 'Amazon Order ID', 'order id', 'orderid',
        'Reference ID', 'Reference'
      ],
      sku: [
        'sku', 'SKU', 'Product SKU', 'Item SKU', 'product-sku',
        'ASIN', 'asin', 'Item ID', 'item-id', 'product_sku',
        'FNSKU', 'MSKU'
      ],
      productName: [
        'product-name', 'Product Name', 'Title', 'Item Name', 'product_name',
        'title', 'item-title', 'Product Title', 'item name', 'product name'
      ],
      quantity: [
        'quantity', 'Quantity', 'Qty', 'qty', 'Quantity Purchased',
        'quantity-purchased', 'Quantity Ordered', 'Units', 'units'
      ],
      unitPrice: [
        'unit-price', 'Unit Price', 'Price', 'Item Price', 'unit_price',
        'price', 'item-price', 'Product Price', 'unit price', 'item price'
      ],
      totalAmount: [
        'total', 'Total', 'Amount', 'Order Total', 'total_amount',
        'total-price', 'Total Price', 'Line Total', 'order-total', 'total amount'
      ],
      fees: [
        'fees', 'Fees', 'Commission', 'Platform Fee', 'fees',
        'amazon-fee', 'Amazon Fee', 'referral-fee', 'Referral Fee', 'commission'
      ],
      netAmount: [
        'net', 'Net Amount', 'Net', 'Payout', 'net_amount',
        'net-amount', 'Net Total', 'Proceeds', 'net amount'
      ],
      orderDate: [
        'order-date', 'Order Date', 'Date', 'Purchase Date', 'order_date',
        'purchase-date', 'Transaction Date', 'Sale Date', 'order date', 'date',
        'Update Date and Time'
      ],
      status: [
        'status', 'Status', 'Order Status', 'State', 'order_status',
        'order-status', 'Fulfillment Status', 'Ship Status', 'order status',
        'Disposition', 'Event Type'
      ],
    };

    // Amazon-specific mapping based on your CSV headers
    if (marketplace === 'amazon') {
      return {
        ...baseMapping,
        // Your CSV has these exact headers: Date, FNSKU, ASIN, MSKU, Title, Event Type, Reference ID, Quantity, Fulfillment Center, Disposition, Reason, Country, Reconciled, Unreconciled, Update Date and Time
        orderId: ['Reference ID', 'Reference', ...baseMapping.orderId],
        sku: ['FNSKU', 'ASIN', 'MSKU', ...baseMapping.sku],
        productName: ['Title', ...baseMapping.productName],
        quantity: ['Quantity', ...baseMapping.quantity],
        orderDate: ['Date', 'Update Date and Time', ...baseMapping.orderDate],
        status: ['Event Type', 'Disposition', ...baseMapping.status],
        // Additional Amazon-specific fields
        fulfillmentCenter: ['Fulfillment Center'],
        reason: ['Reason'],
        country: ['Country'],
        reconciled: ['Reconciled'],
        unreconciled: ['Unreconciled']
      };
    }

    if (marketplace === 'ebay') {
      return {
        ...baseMapping,
        sku: [...baseMapping.sku, 'Item ID', 'Custom Label', 'eBay Item ID'],
        orderId: [...baseMapping.orderId, 'Sales Record Number', 'Transaction ID'],
        fees: [...baseMapping.fees, 'eBay Fee', 'Final Value Fee', 'Insertion Fee'],
      };
    }

    if (marketplace === 'shopify') {
      return {
        ...baseMapping,
        sku: [...baseMapping.sku, 'Variant SKU', 'Product SKU', 'Variant ID'],
        orderId: [...baseMapping.orderId, 'Name', 'Order', 'Order Number'],
        orderDate: [...baseMapping.orderDate, 'Created at', 'Created At'],
        status: [...baseMapping.status, 'Fulfillment Status', 'Financial Status'],
      };
    }

    return baseMapping;
  }

  private storeProcessedRecords(records: SalesRecord[]): void {
    // In a real application, this would store to a database
    const existingRecords = JSON.parse(localStorage.getItem('salesRecords') || '[]');
    const updatedRecords = [...existingRecords, ...records];
    localStorage.setItem('salesRecords', JSON.stringify(updatedRecords));
  }

  public getProcessedRecords(): SalesRecord[] {
    return JSON.parse(localStorage.getItem('salesRecords') || '[]');
  }

  public clearProcessedRecords(): void {
    localStorage.removeItem('salesRecords');
  }
}

File: backend/services/data_processor.py
import pandas as pd
import numpy as np
import logging
from typing import List, Dict, Any, Optional
from datetime import datetime, timedelta
import io
import asyncio
from sqlalchemy.orm import Session
from fastapi import UploadFile

from database.models import SalesRecord, Product, DataProcessingJob, SKUMapping
from services.sku_mapper import SKUMapperService
from utils.logger import setup_logger

logger = setup_logger(__name__)

class DataProcessorService:
    """Comprehensive data processing service for sales data from multiple marketplaces"""
    
    def __init__(self):
        self.sku_mapper = SKUMapperService()
        self.supported_formats = ['.csv', '.xlsx', '.xls']
        
        # Marketplace-specific column mappings
        self.column_mappings = {
            'amazon': {
                'order_id': ['order-id', 'Order ID', 'OrderId', 'order_id', 'Amazon Order ID'],
                'sku': ['sku', 'SKU', 'ASIN', 'Product SKU', 'Item SKU'],
                'product_name': ['product-name', 'Product Name', 'Title', 'Item Name'],
                'quantity': ['quantity', 'Quantity', 'Qty', 'Quantity Purchased'],
                'unit_price': ['unit-price', 'Unit Price', 'Price', 'Item Price'],
                'total_amount': ['total', 'Total', 'Amount', 'Order Total'],
                'fees': ['fees', 'Fees', 'Amazon Fee', 'Referral Fee', 'Commission'],
                'order_date': ['order-date', 'Order Date', 'Purchase Date', 'Date'],
                'status': ['status', 'Status', 'Order Status', 'Fulfillment Status']
            },
            'ebay': {
                'order_id': ['Sales Record Number', 'Order ID', 'Transaction ID'],
                'sku': ['Item ID', 'Custom Label', 'SKU', 'Product ID'],
                'product_name': ['Title', 'Item Title', 'Product Name'],
                'quantity': ['Quantity', 'Qty', 'Quantity Sold'],
                'unit_price': ['Price', 'Sale Price', 'Unit Price'],
                'total_amount': ['Total', 'Sale Amount', 'Total Price'],
                'fees': ['eBay Fee', 'Final Value Fee', 'Fees'],
                'order_date': ['Sale Date', 'Order Date', 'Transaction Date'],
                'status': ['Order Status', 'Status']
            },
            'shopify': {
                'order_id': ['Name', 'Order', 'Order Number', 'Order ID'],
                'sku': ['SKU', 'Variant SKU', 'Product SKU'],
                'product_name': ['Name', 'Product Name', 'Title'],
                'quantity': ['Quantity', 'Qty'],
                'unit_price': ['Price', 'Unit Price', 'Product Price'],
                'total_amount': ['Total', 'Line Total', 'Amount'],
                'fees': ['Fees', 'Transaction Fee', 'Payment Fee'],
                'order_date': ['Created at', 'Order Date', 'Date'],
                'status': ['Fulfillment Status', 'Status', 'Order Status']
            },
            'walmart': {
                'order_id': ['Purchase Order', 'PO Number', 'Order ID'],
                'sku': ['SKU', 'Product SKU', 'Item SKU', 'WM SKU'],
                'product_name': ['Product Name', 'Item Name', 'Title'],
                'quantity': ['Quantity', 'Qty', 'Units'],
                'unit_price': ['Unit Price', 'Price', 'Item Price'],
                'total_amount': ['Total', 'Amount', 'Line Total'],
                'fees': ['Commission', 'Walmart Fee', 'Fees'],
                'order_date': ['Order Date', 'Date', 'Purchase Date'],
                'status': ['Status', 'Order Status', 'Fulfillment Status']
            }
        }
    
    async def process_file(self, file: UploadFile, marketplace: str, job_id: str, db: Session) -> Dict:
        """Process uploaded sales data file"""
        try:
            logger.info(f"Starting file processing: {file.filename} for {marketplace}")
            
            # Update job status
            job = db.query(DataProcessingJob).filter(DataProcessingJob.id == job_id).first()
            if not job:
                raise Exception("Processing job not found")
            
            job.status = "processing"
            db.commit()
            
            # Read file content
            content = await file.read()
            df = await self._read_file_content(content, file.filename)
            
            job.records_total = len(df)
            db.commit()
            
            # Process data
            processed_records = []
            errors = []
            warnings = []
            
            for index, row in df.iterrows():
                try:
                    # Map columns to standard format
                    mapped_data = self._map_columns(row.to_dict(), marketplace)
                    
                    # Validate required fields
                    validation_result = self._validate_record(mapped_data)
                    if not validation_result['valid']:
                        errors.extend([f"Row {index + 1}: {error}" for error in validation_result['errors']])
                        continue
                    
                    # Process SKU mapping
                    sku_mapping = self.sku_mapper.map_sku(mapped_data['sku'], marketplace, db)
                    
                    # Create sales record
                    sales_record = self._create_sales_record(mapped_data, sku_mapping, marketplace)
                    processed_records.append(sales_record)
                    
                    # Update progress
                    job.records_processed = index + 1
                    job.progress = int((index + 1) / len(df) * 100)
                    
                    if index % 10 == 0:  # Update every 10 records
                        db.commit()
                    
                except Exception as e:
                    error_msg = f"Row {index + 1}: {str(e)}"
                    errors.append(error_msg)
                    logger.error(error_msg)
            
            # Save processed records to database
            await self._save_records_to_db(processed_records, db)
            
            # Update job completion
            job.status = "completed"
            job.completed_at = datetime.utcnow()
            job.progress = 100
            job.errors = errors
            job.warnings = warnings
            db.commit()
            
            result = {
                'job_id': job_id,
                'records_processed': len(processed_records),
                'records_total': len(df),
                'errors': errors,
                'warnings': warnings,
                'success_rate': len(processed_records) / len(df) if len(df) > 0 else 0
            }
            
            logger.info(f"File processing completed: {result}")
            return result
            
        except Exception as e:
            logger.error(f"Error processing file: {str(e)}")
            
            # Update job with error status
            if 'job' in locals():
                job.status = "failed"
                job.errors = [str(e)]
                db.commit()
            
            raise
    
    async def _read_file_content(self, content: bytes, filename: str) -> pd.DataFrame:
        """Read file content into pandas DataFrame"""
        try:
            if filename.endswith('.csv'):
                # Try different encodings
                for encoding in ['utf-8', 'latin-1', 'cp1252']:
                    try:
                        df = pd.read_csv(io.BytesIO(content), encoding=encoding)
                        break
                    except UnicodeDecodeError:
                        continue
                else:
                    raise Exception("Unable to decode CSV file with supported encodings")
            
            elif filename.endswith(('.xlsx', '.xls')):
                df = pd.read_excel(io.BytesIO(content))
            
            else:
                raise Exception(f"Unsupported file format: {filename}")
            
            # Clean column names
            df.columns = df.columns.str.strip()
            
            # Remove empty rows
            df = df.dropna(how='all')
            
            logger.info(f"Successfully read file: {len(df)} rows, {len(df.columns)} columns")
            return df
            
        except Exception as e:
            logger.error(f"Error reading file content: {str(e)}")
            raise
    
    def _map_columns(self, row_data: Dict, marketplace: str) -> Dict:
        """Map marketplace-specific columns to standard format"""
        mapped_data = {}
        column_mapping = self.column_mappings.get(marketplace, {})
        
        for standard_field, possible_columns in column_mapping.items():
            value = None
            
            # Find the first matching column
            for col_name in possible_columns:
                if col_name in row_data and pd.notna(row_data[col_name]):
                    value = row_data[col_name]
                    break
            
            mapped_data[standard_field] = value
        
        # Add any unmapped columns as additional data
        mapped_data['raw_data'] = row_data
        
        return mapped_data
    
    def _validate_record(self, record: Dict) -> Dict:
        """Validate required fields in sales record"""
        errors = []
        
        # Required fields
        required_fields = ['order_id', 'sku', 'product_name', 'quantity', 'unit_price']
        
        for field in required_fields:
            if not record.get(field) or pd.isna(record[field]):
                errors.append(f"Missing required field: {field}")
        
        # Validate data types
        try:
            if record.get('quantity'):
                quantity = float(record['quantity'])
                if quantity <= 0:
                    errors.append("Quantity must be greater than 0")
        except (ValueError, TypeError):
            errors.append("Invalid quantity format")
        
        try:
            if record.get('unit_price'):
                price = float(record['unit_price'])
                if price < 0:
                    errors.append("Unit price cannot be negative")
        except (ValueError, TypeError):
            errors.append("Invalid unit price format")
        
        # Validate order date
        if record.get('order_date'):
            try:
                pd.to_datetime(record['order_date'])
            except:
                errors.append("Invalid order date format")
        
        return {
            'valid': len(errors) == 0,
            'errors': errors
        }
    
    def _create_sales_record(self, mapped_data: Dict, sku_mapping: Dict, marketplace: str) -> SalesRecord:
        """Create SalesRecord object from mapped data"""
        # Calculate derived fields
        quantity = int(float(mapped_data.get('quantity', 1)))
        unit_price = float(mapped_data.get('unit_price', 0))
        total_amount = float(mapped_data.get('total_amount', 0))
        fees = float(mapped_data.get('fees', 0))
        
        # Calculate total if not provided
        if not total_amount and unit_price and quantity:
            total_amount = unit_price * quantity
        
        # Calculate net amount
        net_amount = total_amount - fees
        
        # Parse order date
        order_date = datetime.utcnow()
        if mapped_data.get('order_date'):
            try:
                order_date = pd.to_datetime(mapped_data['order_date']).to_pydatetime()
            except:
                pass
        
        sales_record = SalesRecord(
            order_id=str(mapped_data['order_id']),
            marketplace=marketplace,
            sku=sku_mapping['original_sku'],
            msku=sku_mapping.get('msku'),
            product_id=sku_mapping.get('product_id'),
            product_name=str(mapped_data['product_name']),
            quantity=quantity,
            unit_price=unit_price,
            total_amount=total_amount,
            fees=fees,
            net_amount=net_amount,
            order_date=order_date,
            status=mapped_data.get('status', 'pending'),
            customer_name=mapped_data.get('customer_name'),
            customer_email=mapped_data.get('customer_email'),
            customer_address=mapped_data.get('customer_address')
        )
        
        return sales_record
    
    async def _save_records_to_db(self, records: List[SalesRecord], db: Session):
        """Save processed records to database"""
        try:
            # Save in batches for better performance
            batch_size = 100
            for i in range(0, len(records), batch_size):
                batch = records[i:i + batch_size]
                db.add_all(batch)
                db.commit()
                
                # Small delay to prevent overwhelming the database
                await asyncio.sleep(0.01)
            
            logger.info(f"Successfully saved {len(records)} records to database")
            
        except Exception as e:
            logger.error(f"Error saving records to database: {str(e)}")
            db.rollback()
            raise
    
    def get_processing_statistics(self, db: Session, days: int = 30) -> Dict:
        """Get data processing statistics"""
        try:
            start_date = datetime.utcnow() - timedelta(days=days)
            
            jobs = db.query(DataProcessingJob).filter(
                DataProcessingJob.created_at >= start_date
            ).all()
            
            total_jobs = len(jobs)
            completed_jobs = len([j for j in jobs if j.status == 'completed'])
            failed_jobs = len([j for j in jobs if j.status == 'failed'])
            total_records = sum(j.records_processed for j in jobs)
            
            marketplace_stats = {}
            for job in jobs:
                if job.marketplace not in marketplace_stats:
                    marketplace_stats[job.marketplace] = {
                        'jobs': 0,
                        'records': 0,
                        'success_rate': 0
                    }
                marketplace_stats[job.marketplace]['jobs'] += 1
                marketplace_stats[job.marketplace]['records'] += job.records_processed
            
            return {
                'total_jobs': total_jobs,
                'completed_jobs': completed_jobs,
                'failed_jobs': failed_jobs,
                'success_rate': completed_jobs / total_jobs if total_jobs > 0 else 0,
                'total_records_processed': total_records,
                'marketplace_breakdown': marketplace_stats,
                'period_days': days
            }
            
        except Exception as e:
            logger.error(f"Error getting processing statistics: {str(e)}")
            raise
    
    def cleanup_old_jobs(self, db: Session, days: int = 90):
        """Clean up old processing jobs"""
        try:
            cutoff_date = datetime.utcnow() - timedelta(days=days)
            
            old_jobs = db.query(DataProcessingJob).filter(
                DataProcessingJob.created_at < cutoff_date,
                DataProcessingJob.status.in_(['completed', 'failed'])
            ).all()
            
            for job in old_jobs:
                db.delete(job)
            
            db.commit()
            logger.info(f"Cleaned up {len(old_jobs)} old processing jobs")
            
        except Exception as e:
            logger.error(f"Error cleaning up old jobs: {str(e)}")
            db.rollback()
            raise